### PR TITLE
Add 32 bit compressed pointer support.

### PR DESCRIPTION
**Pull request by zherczeg**
This is a WIP patch, which will not be landed before the 1.0 release.

This patch extends compressed pointers to be 32 bit long when "--cpointer_32_bit on" is passed. After custom allocators will be added to Jerry, this will allow to use the full 32 bit address space on 32 bit systems at the cost of increased memory size. I will do some measurements to estimate this trade-of with the current allocator.

### CODE REVIEW

**Comment by zherczeg** [comment]
Around 20% memory increase (9% RSS increase), a little perf gain (2%) and a small binary size reduction (200 byte).

Benchmark | Heap (bytes) | RSS (bytes) | Perf (sec) |
----: | ----: | ----: | ----: | 
3d-cube.js | 16376 -> 24568 : -50.024% | 57344 -> 65536 : -14.286% | 0.944 -> 0.929 : +1.550% | 
3d-raytrace.js | 161152 -> 197488 : -22.548% | 217088 -> 212992 : +1.887% | 1.116 -> 1.157 : -3.704% | 
access-binary-trees.js | 32760 -> 49136 : -49.988% | 53248 -> 69632 : -30.769% | 0.592 -> 0.549 : +7.289% | 
access-fannkuch.js | 7608 -> 8184 : -7.571% | 24576 -> 24576 : 0.000% | 2.363 -> 2.330 : +1.409% | 
access-nbody.js | 6688 -> 8176 : -22.249% | 28672 -> 28672 : 0.000% | 1.077 -> 1.058 : +1.752% | 
bitops-3bit-bits-in-byte.js | 928 -> 1072 : -15.517% | 16384 -> 20480 : -25.000% | 0.591 -> 0.579 : +1.951% | 
bitops-bits-in-byte.js | 880 -> 1032 : -17.273% | 16384 -> 20480 : -25.000% | 0.858 -> 0.853 : +0.591% | 
bitops-bitwise-and.js | 568 -> 664 : -16.901% | 16384 -> 16384 : 0.000% | 0.934 -> 0.883 : +5.476% | 
bitops-nsieve-bits.js | 138568 -> 175248 : -26.471% | 155648 -> 192512 : -23.684% | 1.788 -> 1.774 : +0.764% | 
controlflow-recursive.js | 912 -> 1048 : -14.912% | 94208 -> 98304 : -4.348% | 0.399 -> 0.388 : +2.775% | 
crypto-aes.js | 49176 -> 57376 : -16.675% | 94208 -> 102400 : -8.696% | 1.113 -> 1.096 : +1.502% | 
crypto-md5.js | 139256 -> 172024 : -23.531% | 163840 -> 208896 : -27.500% | 0.738 -> 0.751 : -1.886% | 
crypto-sha1.js | 90104 -> 106488 : -18.183% | 110592 -> 126976 : -14.815% | 0.705 -> 0.704 : +0.008% | 
date-format-tofte.js | 16376 -> 16376 : 0.000% | 40960 -> 40960 : 0.000% | 0.864 -> 0.841 : +2.626% | 
date-format-xparb.js | 16384 -> 24576 : -50.000% | 49152 -> 53248 : -8.333% | 0.470 -> 0.440 : +6.270% | 
math-cordic.js | 2128 -> 2640 : -24.060% | 20480 -> 20480 : 0.000% | 1.345 -> 1.302 : +3.180% | 
math-partial-sums.js | 1432 -> 1776 : -24.022% | 16384 -> 20480 : -25.000% | 0.708 -> 0.695 : +1.804% | 
math-spectral-norm.js | 8184 -> 8176 : +0.098% | 24576 -> 24576 : 0.000% | 0.599 -> 0.581 : +3.107% | 
string-base64.js | 92032 -> 93472 : -1.565% | 143360 -> 139264 : +2.857% | 2.054 -> 1.991 : +3.047% | 
string-fasta.js | 16320 -> 16368 : -0.294% | 32768 -> 32768 : 0.000% | 1.402 -> 1.338 : +4.586% | 
Geometric mean: | -19.174% | -9.542% | +2.237% | 

Binary sizes (bytes)
314e74f8ce:156952
5600c4c206:156724

**Comment by LaszloLango** [comment]
> Enable cpointer_32_bit by default (this will be removed after testing).

IMHO, we should add a new travis job and test this feature on every new commit.

**Comment by zherczeg** [comment]
As you can see travis is failing, since you need a 32 bit system (travis is 64 bit by default). I dont have any idea how to test it.

File: jerry-core/ecma/base/ecma-globals.h [hunk 1]
```diff
@@ -361,9 +361,16 @@ typedef struct
...
+  jmem_cpointer_t next_property_cp; /**< next cpointer */
+#endif /* JERRY_CPOINTER_32_BIT */
   ecma_property_t types[ECMA_PROPERTY_PAIR_ITEM_COUNT]; /**< two property type slot. The first represent
                                                          *   the type of this property (e.g. property pair) */
+#ifdef JERRY_CPOINTER_32_BIT
+  uint16_t padding; /**< an unused value */
+#else /* !JERRY_CPOINTER_32_BIT */
   jmem_cpointer_t next_property_cp; /**< next cpointer */
+#endif /* JERRY_CPOINTER_32_BIT */
 } ecma_property_header_t;
```
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-globals.h]: Can you explain this change? Why did you move the next cpointer on 32 bit?

Comment (zherczeg) [inline jerry-core/ecma/base/ecma-globals.h]: Because ECMA_PROPERTY_VALUE_PTR expects that the type fields are packed in a 4 byte word, which is followed by maximum of four value words. For 16 bit cpointers the next pointer can be squeezed after the last two bytes of the first word. This is not possible with 32 bit cpointers, where that 16 bit is unused.

**Comment by LaszloLango** [comment]
The code looks good, but `docs/04.INTERNALS.md` should be updated.

**Comment by zherczeg** [comment]
Description added, patch updated.

**Comment by LaszloLango** [comment]
LGTM

File: jerry-core/config.h [hunk 2]
```diff
@@ -39,8 +39,6 @@
  */
 #ifndef CONFIG_MEM_HEAP_AREA_SIZE
 # define CONFIG_MEM_HEAP_AREA_SIZE (512 * 1024)
-#elif CONFIG_MEM_HEAP_AREA_SIZE > (512 * 1024)
-# error "Currently, maximum 512 kilobytes heap size is supported"
```
Comment (dbatyai) [inline jerry-core/config.h]: Removing this may be a bad idea, I think it would be better to add `&& !JERRY_CPOINTER_32_BIT` to the condition instead.

**Comment by dbatyai** [comment]
LGTM otherwise.

File: jerry-core/jmem/jmem-heap.c [hunk 3]
```diff
@@ -145,8 +145,10 @@ static void jmem_heap_stat_free_iter ();
 void
 jmem_heap_init (void)
 {
-  JERRY_STATIC_ASSERT ((1u << JMEM_HEAP_OFFSET_LOG) >= JMEM_HEAP_SIZE,
-                       two_pow_mem_heap_offset_should_not_be_less_than_mem_heap_size);
+#ifndef JERRY_CPOINTER_32_BIT
+  JERRY_STATIC_ASSERT (((UINT16_MAX + 1) << JMEM_ALIGNMENT_LOG) >= JMEM_HEAP_SIZE,
+                       maximum_heap_size_for_16_bit_compressed_pointers_is_512K);
```
Comment (zherczeg) [inline jerry-core/jmem/jmem-heap.c]: Here.

Comment (dbatyai) [inline jerry-core/jmem/jmem-heap.c]: Oh, I see. Okay.

**Comment by dbatyai** [comment]
LGTM
